gc/base/SparseVirtualMemory.hpp

```diff
@@ -0,0 +1,173 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2022 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/**
+ * @file
+ * @ingroup GC_Base_Core
+ */
+
+#if !defined(SPARSEVIRTUALMEMORY_HPP_)
+#define SPARSEVIRTUALMEMORY_HPP_
+
+#include "omrcomp.h"
+#include "omrport.h"
+#include "modronbase.h"
+#include "modronopt.h"
+
+#include "BaseVirtual.hpp"
+#include "VirtualMemory.hpp"
+
+class MM_GCExtensions;
+class MM_GCExtensionsBase;
+class MM_Heap;
+class MM_SparseAddressOrderedFixedSizeDataPool;
+struct J9PortVmemParams;
+
+/**
+ * Large virtual memory for allocation of large objects (their data portion). It's sparsely
+ * committed only for live objects (memory is eagerly committed/de-committed on allocate/free).
+ *
+ * It also maintains mapping from the data portion to the in-heap object pointer.
+ * The reverse mapping is not maintained (it's expected that the heap object has an easy way to
+ * find the data portion, for example, via the dataAddr pointer in the object header).
+ *
+ * @ingroup GC_Base_Core
+ */
+class MM_SparseVirtualMemory : public MM_VirtualMemory {
+/*
+ * Data members
+ */
+private:
+	MM_Heap *_heap; /**< reference to in-heap */
+	MM_SparseAddressOrderedFixedSizeDataPool *_sparseDataPool; /**< Structure that manages data and free region of sparse virtual memory */
+	omrthread_monitor_t _largeObjectVirtualMemoryMutex;
+protected:
+public:
+/*
+ * Function members
+ */
+private:
+
+	bool initialize(MM_EnvironmentBase* env, uint32_t memoryCategory);
+	void tearDown(MM_EnvironmentBase *env);
+	void kill(MM_EnvironmentBase *env);
```

Comment:
The same comment as above for `MM_SparseAddressOrderedFixedSizeDataPool`
- should `initialize()` and `tearDown()` be `protected`?
- should `kill()` be `public` and be called when we tear down class called `MM_SparseVirtualMemory::newInstance()`?

Reply:
Yes, thank you, good point. Will make these changes.